Boundless reads that don't alter pixels

sgillies opened the pull request:
I'm proposing a new test of boundless reads. The test asserts that a boundless read with no resampling (nearest neighbor, in other words) from our testing dataset gives the same results as a not-boundless read. For example, do boundless read with a window 1 pixel larger on every side. We expect that the interior of this read would be the same as a not-boundless read.

The `WarpedVRT` usage change is adapted from #1238. @ungarj: could you check this out and let me know if it matches your expectations?

sgillies commented on tests/test_boundless_read.py:
```diff
@@ -0,0 +1,15 @@
+from rasterio.windows import Window
+
+
+def test_pixel_fidelity(path_rgb_byte_tif):
```
TODO: vary the size of the buffer.

sgillies commented on tests/test_boundless_read.py:
```diff
@@ -0,0 +1,19 @@
+from rasterio.windows import Window
+
+
+@given(st.integers(min_value=0, max_value=700))
```
I'd go higher on the size, but this isn't about testing memory errors.

sgillies commented on tests/test_boundless_read.py:
```diff
@@ -0,0 +1,19 @@
+        rgb = dataset.read()
+        rgb_padded = dataset.read(window=Window(-size, -size, dataset.width + 2 * size, dataset.height + 2 * size), boundless=True)
+        assert rgb_padded.shape == (3, dataset.height + 2 * size, dataset.width + 2 * size)
+    assert numpy.all(rgb == rgb_padded[:, size:(-size or None), size:(-size or None)])
```
`None` is what we want instead of `0` at the end of a slice.

brendan-ward replied:
You may want to add that as a comment, it isn't immediately obvious.

brendan-ward commented on tests/test_boundless_read.py:
```diff
@@ -0,0 +1,19 @@
+@given(st.integers(min_value=0, max_value=700))
+def test_outer_boundless_pixel_fidelity(size):
+    """An outer boundless read doesn't change pixels"""
+    path_rgb_byte_tif = 'tests/data/RGB.byte.tif'
```
Note `path_rgb_byte_tif` is a fixture available from `conftest.py`, no need to define it here.

sgillies replied:
I thought I remembered that pytest fixtures and hypothesis givens don't play well together, but I've figured it out.

brendan-ward commented on tests/test_boundless_read.py:
```diff
@@ -0,0 +1,19 @@
+    path_rgb_byte_tif = 'tests/data/RGB.byte.tif'
+    with rasterio.open(path_rgb_byte_tif) as dataset:
+        rgb = dataset.read()
+        rgb_padded = dataset.read(window=Window(-size, -size, dataset.width + 2 * size, dataset.height + 2 * size), boundless=True)
```
Declare variables for `padded_width = dataset.width + 2 * w_padding` and likewise for height, and use here and in the assertion.

brendan-ward commented on tests/test_boundless_read.py:
```diff
@@ -0,0 +1,19 @@
+    with rasterio.open(path_rgb_byte_tif) as dataset:
+        rgb = dataset.read()
+        rgb_padded = dataset.read(window=Window(-size, -size, dataset.width + 2 * size, dataset.height + 2 * size), boundless=True)
+        assert rgb_padded.shape == (3, dataset.height + 2 * size, dataset.width + 2 * size)
```
`dataset.count` instead of `3`

brendan-ward replied:
What about just testing the first band?  It would simplify a little bit, and you aren't really testing multi-band behavior here, only boundless reads.

brendan-ward commented on tests/test_boundless_read.py:
```diff
@@ -0,0 +1,19 @@
+
+
+@given(st.integers(min_value=0, max_value=700))
+def test_outer_boundless_pixel_fidelity(size):
```
Of course, if you wanted to leverage `hypothesis`, padding all 4 sides independently would be the way to go.  Then no matter how any side is padded, we can be sure that the data within the padding is identical from a boundless read to a non boundless read.

sgillies commented:
Indeed, `resampling` was missing from the docstrings. I've added it and also added a note about the existence of the dataset's block cache and that resampling is downstream of the cache.

ungarj commented:
@sgillies I just checked and it now works as expected. Also, thanks for updating the docstrings!
